pymc3/math.py

```diff
@@ -136,3 +142,95 @@ def expand_packed_triangular(n, packed, lower=True, diagonal_only=False):
+    """
+    __props__ = ()
+
+    def make_node(self, diag):
+        diag = tt.as_tensor_variable(diag)
+        if diag.type.ndim != 2:
+            raise TypeError('data argument must be a matrix', diag.type)
+
+        return tt.Apply(self, [diag], [tt.tensor3(dtype=diag.dtype)])
+
+    def perform(self, node, ins, outs, params=None):
+        (C,) = ins
+        (z,) = outs
+
+        bc = C.shape[0]
+        dim = C.shape[-1]
+        Cd = np.zeros((bc, dim, dim), C.dtype)
+        bidx = np.repeat(np.arange(bc), dim)
+        didx = np.tile(np.arange(dim), bc)
+        Cd[bidx, didx, didx] = C.flatten()
+        z[0] = Cd
+
+    def grad(self, inputs, gout):
+        (gz,) = gout
+        idx = tt.arange(gz.shape[-1])
+        return [gz[..., idx, idx]]
+
+    def infer_shape(self, nodes, shapes):
+        return [(shapes[0][0], ) + (shapes[0][1],) * 2]
+
+
+def batched_diag(C):
+    C = tt.as_tensor(C)
+    dim = C.shape[-1]
+    if C.ndim == 2:
+        # diag -> matrices
+        return BatchedDiag()(C)
+    elif C.ndim == 3:
+        # matrices -> diag
+        idx = tt.arange(dim)
+        return C[..., idx, idx]
+    else:
+        raise ValueError('Input should be 2 or 3 dimensional')
+
+
+class BlockDiagonalMatrix(Op):
+    __props__ = ('sparse', 'format')
+
+    def __init__(self, sparse=False, format='csr'):
+        if format not in ('csr', 'csc'):
+            raise ValueError("format must be one of: 'csr', 'csc'", format)
+        self.sparse = sparse
+        self.format = format
+
+    def make_node(self, *matrices):
+        if not matrices:
+            raise ValueError('Got no matrices to allocate')
+        matrices = list(map(tt.as_tensor, matrices))
+        if any(mat.type.ndim != 2 for mat in matrices):
+            raise TypeError('all data arguments must be matrices')
+        if self.sparse:
+            out_type = theano.sparse.matrix(self.format, dtype=largest_common_dtype(matrices))
+        else:
+            out_type = theano.tensor.matrix(dtype=largest_common_dtype(matrices))
+        return tt.Apply(self, matrices, [out_type])
+
+    def perform(self, node, inputs, output_storage, params=None):
+        dtype = largest_common_dtype(inputs)
+        if self.sparse:
+            output_storage[0][0] = sp.sparse.block_diag(
+                inputs, self.format, dtype
+            )
+        else:
+            output_storage[0][0] = scipy_block_diag(*inputs).astype(dtype)
+
+    def grad(self, inputs, gout):
+        shapes = tt.stack([i.shape for i in inputs])
+        index_end = shapes.cumsum(0)
+        index_begin = index_end - shapes
+        slices = [ix_(tt.arange(index_begin[i, 0], index_end[i, 0]),
+                      tt.arange(index_begin[i, 1], index_end[i, 1])
+                      ) for i in range(len(inputs))]
+        return [gout[0][slc] for slc in slices]
+
+    def infer_shape(self, nodes, shapes):
+        first, second = zip(*shapes)
+        return [(tt.add(*first), tt.add(*second))]
```

Comment:
@nouiz How do I write an optimizer in case I have only one input? I tried this code but it did not work
```python
block_diagonal = BlockDiagonalMatrix()

_simplify_single_block_diagonal = theano.gof.PatternSub(
    (block_diagonal, 'x'), 'x')
simplify_single_block_diagonal = theano.gof.NavigatorOptimizer(_simplify_single_block_diagonal)
theano.compile.optdb.register('simplify_single_block_diagonal', simplify_single_block_diagonal, 'last', 'fast_run')
```